### PR TITLE
✨Started a config parser for the amp-experiment 1.0 format

### PR DESCRIPTION
**Pull request by torch2424**
relates to #20225
relates to #21484

This starts the config parser for the 1.0 format (see #20225). Most importantly here, this preserves the analytics, but also passes an object down to `applyMutations_()` which can then use a mutation service to start ingestion of mutations (later PR).

* [x] **This still requires tests. Want code approved before writing the tests** 😄 

# Example

**First image had some testing code to show things work**

Screen Shot 2019-03-14 at 11 39 18 AM

Screen Shot 2019-03-21 at 2 31 45 PM

### CODE REVIEW

File: extensions/amp-experiment/1.0/variant.js [hunk 1]
```diff
@@ -42,12 +41,12 @@ export class Variants {
...
 
   /**
-   * @param {!Object<string, ?string>|!Promise<!Object<string, ?string>>} variants
+   * @param {!Promise<!Object<string, ?string>>} variants
    * @package
    * @restricted
    */
   init(variants) {
-    this.variantsDeferred_.resolve(variants);
+    variants.then(this.variantsDeferred_.resolve);
```
Comment (zhouyx) [inline extensions/amp-experiment/1.0/variant.js]: I believe you want to do 
`variants.then(result => this.viriantsDeferred_.resolve(result))` ?

File: extensions/amp-experiment/1.0/variant.js [hunk 2]
```diff
@@ -119,7 +117,7 @@ export function allocateVariant(ampdoc, experimentName, config) {
           // enumeration is implementation (browser) dependent.
           const variantNames = Object.keys(config['variants']).sort();
           for (let i = 0; i < variantNames.length; i++) {
-            upperBound += config['variants'][variantNames[i]];
+            upperBound += config['variants'][variantNames[i]].weight;
```
Comment (zhouyx) [inline extensions/amp-experiment/1.0/variant.js]: config is a JsonObject. We will need to use `['weight']` : ) Singlepass~

File: extensions/amp-experiment/1.0/variant.js [hunk 3]
```diff
@@ -131,25 +129,36 @@ export function allocateVariant(ampdoc, experimentName, config) {
...
-      const percentage = variants[variantName];
+
+      const variant = variants[variantName];
+      assertVariant(
+          variant,
+          experimentName,
+          variantName
+      );
+
+      const percentage = variant.weight;
```
Comment (zhouyx) [inline extensions/amp-experiment/1.0/variant.js]: nit: variant['weight']

File: extensions/amp-experiment/1.0/variant.js [hunk 4]
```diff
@@ -131,25 +129,36 @@ export function allocateVariant(ampdoc, experimentName, config) {
...
+
       userAssert(
-          typeof percentage === 'number' && percentage > 0 && percentage < 100,
+          percentage && percentage > 0 && percentage < 100,
```
Comment (zhouyx) [inline extensions/amp-experiment/1.0/variant.js]: can we move the assert to `#assertVariant()` as well?

File: extensions/amp-experiment/1.0/variant.js [hunk 5]
```diff
@@ -191,3 +200,30 @@ function assertName(name) {
...
+
+  // Assert the variant weight
+  userAssert(
+      variant['weight'] !== undefined &&
+    typeof variant['weight'] === 'number',
+      `${experimentName}.${variantName} must have a weight.`);
+
+  // Assert the variant mutations
+  userAssert(
+      isArray(variant['mutations']),
```
Comment (zhouyx) [inline extensions/amp-experiment/1.0/variant.js]: what will be included in the `mutations`? Is it still required if it is an empty array?

File: extensions/amp-experiment/1.0/amp-experiment.js [hunk 6]
```diff
@@ -75,23 +76,48 @@ export class AmpExperiment extends AMP.BaseElement {
...
    * @private
    */
-  addToBody_(experiments) {
+  applyExperimentVariants_(config, experiments) {
+
+    const appliedExperimentVariantPromises = [];
+
+    for (const name in experiments) {
+      if (experiments[name]) {
+        const variantObject = config[name]['variants'][experiments[name]];
```
Comment (zhouyx) [inline extensions/amp-experiment/1.0/amp-experiment.js]: sigh.. 👿

**Comment by torch2424** [comment]
@zhouyx I have a few questions about the comments you made, so I'll go ahead and wait till Monday, and make all the fixes at once 😄 No worries if you are not in the office Monday, as I could ask them later in the review 👍

**Comment by torch2424** [comment]
@zhouyx Also, when you finish the review, feel free to mention me here so I get the github notification 😂

File: extensions/amp-experiment/1.0/variant.js (hunk 5 above)
Comment (zhouyx) [reply extensions/amp-experiment/1.0/variant.js]: check the length of the array as well

**Comment by torch2424** [comment]
@zhouyx Made all requested changes, and fixed/added appropriate tests 😄 

Please give this another review once you get the chance!

File: extensions/amp-experiment/1.0/test/test-amp-experiment.js [hunk 7]
```diff
@@ -143,12 +149,8 @@ describes.realWin('amp-experiment', {
             'experiment-2': 'variant-d',
             'experiment-3': null,
           });
-          expectBodyHasAttributes({
```
Comment (zhouyx) [inline extensions/amp-experiment/1.0/test/test-amp-experiment.js]: do we still want to test the attributes appended to body?

File: extensions/amp-experiment/1.0/amp-experiment.js [hunk 8]
```diff
@@ -75,23 +92,49 @@ export class AmpExperiment extends AMP.BaseElement {
...
+   * to apply the experiment to the document.
+   * Experiment with no variant assigned (null) will be skipped.
+   * @param {!JsonObject} config
+   * @param {!Object<string, ?string>} experimentToVariant
    * @return {!Promise<!Object<string, ?string>>} a promise of the original
    *     param passed in
    * @private
    */
-  addToBody_(experiments) {
+  applyExperimentVariants_(config, experimentToVariant) {
```
[the remark this answers is not in the document]
Comment (torch2424) [reply extensions/amp-experiment/1.0/amp-experiment.js]: Add comment here on experimentToVariant

File: extensions/amp-experiment/1.0/amp-experiment.js [hunk 9]
```diff
@@ -49,18 +50,34 @@ export class AmpExperiment extends AMP.BaseElement {
             /** @private @const {!Promise<!Object<string, ?string>>} */
             const experimentVariants = Promise.all(variants)
                 .then(() => results)
-                .then(this.addToBody_.bind(this));
+                .then(this.applyExperimentVariants_.bind(this, config));
```
[the remark this answers is not in the document]
Comment (torch2424) [reply extensions/amp-experiment/1.0/amp-experiment.js]: Discussed offline, this does do two :)

File: extensions/amp-experiment/1.0/amp-experiment.js [hunk 10]
```diff
@@ -75,23 +92,49 @@ export class AmpExperiment extends AMP.BaseElement {
...
-          body.setAttribute(ATTR_PREFIX + name,
-              dev().assertString(experiments[name]));
-        }
-      }
-      return experiments;
+      // TODO (torch2424): Use a mutation service,
+      // and apply mutations
+      // Placehodler to pass linting for code review
+      // and keep PRs small
+      body.setAttribute(experimentName, JSON.stringify(variantObject));
```
[the remark this answers is not in the document]
Comment (torch2424) [reply extensions/amp-experiment/1.0/amp-experiment.js]: Discussed offline, make a better NOOP.

File: extensions/amp-experiment/1.0/test/test-amp-experiment.js (hunk 7 above)
Comment (torch2424) [reply extensions/amp-experiment/1.0/test/test-amp-experiment.js]: Nope we dont, the body thing was just a NOOP for linting

**Comment by torch2424** [comment]
@zhouyx Made requested changes / replied to comments! 🎉 

This is good to go 😄

**Comment by zhouyx** [comment]
Look great! Thank you for the change 🎉

**Comment by torch2424** [comment]
@zhouyx Yay! Thanks for the review! 😄 Merging...

**Pull request merged**